var/spack/repos/builtin/packages/adios2/package.py

```diff
@@ -281,3 +283,58 @@ def setup_run_environment(self, env):
             env.prepend_path("HDF5_PLUGIN_PATH", os.path.dirname(all_libs[idx]))
         except ValueError:
             pass
+
+    @run_after("install")
+    def setup_install_tests(self):
+        """
+        Copy the example files after the package is installed to an
+        install test subdirectory for use during `spack test run`.
+        """
+        extra_install_tests = [join_path("testing", "install", "C")]
+        self.cache_extra_test_sources(extra_install_tests)
+
+    def test_run_executables(self):
+        """Run installed adios2 executables"""
+
+        commands_and_args = [("bpls", ["-v", "-V"]), ("adios2-config", ["-v"])]
+
+        for cmd, opts in commands_and_args:
+            with test_part(
+                self,
+                f"test_run_executables_{cmd}",
+                purpose=f"run installed adios2 executable {cmd}",
+            ):
+                exe = which(join_path(self.prefix.bin, cmd))
+                exe(*opts)
+
+    def test_examples(self):
+        """Build and run an example program"""
+        src_dir = self.test_suite.current_test_cache_dir.testing.install.C
+        test_stage_dir = self.test_suite.test_dir_for_spec(self.spec)
+
+        # Create the build tree within this spec's test stage dir so it gets
+        # cleaned up automatically
+        build_dir = tempfile.mkdtemp(dir=test_stage_dir)
```

Comment:
You were too fast for me.  I was trying to post when an update occurred:

BTW. You don't need to create a build subdirectory of the test stage since you're only building tests for one (implicit) test part (i.e., `test_examples`) and your process does not appear to require a subdirectory.  Which means you also don't need to use `with working_dir()` at line 327.

To confirm, the tests run with and without this build subdirectory of the test stage.  As is, the test stage for this spec (`self.test_suite.test_dir_for_spec(self.spec)`) contains two subdirectories:

- `cache`
- `<temp-subdirectory>`

where `<temp-subdirectory>` (after running the test with `spack [-v] test run --keep-stage adios2`) contains:

```
adios_adios2c_test*  CMakeCache.txt       CTestTestfile.cmake 
adios_c_mpi_test*    CMakeFiles/          Makefile 
adios_c_test*        cmake_install.cmake  
```

These are the same contents of `self.test_suite.test_dir_for_spec(self.spec)` if you remove `build_dir`.

Reply:
I don't know what the `cache` subdirectory contains, and maybe it's no problem to have it in my build tree.  But normally I like to build in a clean, empty directory.  Is this suggestion about avoiding leaving a bad example around for others to follow, or is it about performance, or code-cleanliness?

Reply:
The `cache` directory is the one Spack creates when it copies your `setup_install_test` files to the test stage directory.  (And I neglected to mention it is included in the directory if you removed `build_dir`.)  You're accessing it at line 312.

Understandable about wanting a clean, empty directory.  It's more an FYI as I wouldn't call it a "bad" example.

Reply:
I've enabled auto-merge.

Reply:
Sounds good, thanks for your reviews @tldahlgren :grin: